plugins/network/src/tcp.cpp

```diff
@@ -1,77 +1,95 @@
+} // anonymous namespace
+
 // =-=-=-=-=-=-=-
 // local function to read a buffer from a socket
 irods::error tcp_socket_read(
     int             _socket,
     void*           _buffer,
     int             _length,
     int&            _bytes_read,
-    struct timeval* _time_value ) {
-    // =-=-=-=-=-=-=-
-    // Initialize the file descriptor set
-    fd_set set;
-    FD_ZERO( &set );
-    FD_SET( _socket, &set );
+    struct timeval* _time_value )
+{
+    log_net::debug("{}: BEGIN", __func__);
 
-    // =-=-=-=-=-=-=-
-    // local copy of time value?
-    struct timeval timeout;
-    if ( _time_value != NULL ) {
-        timeout = ( *_time_value );
-    }
-
-    // =-=-=-=-=-=-=-
-    // local working variables
-    int   len_to_read = _length;
-    char* read_ptr    = static_cast<char*>( _buffer );
+    // NOLINTNEXTLINE(bugprone-lambda-function-name)
+    irods::at_scope_exit log_exit{[fn = __func__] { log_net::debug("{}: END", fn); }};
 
-    // =-=-=-=-=-=-=-
-    // reset bytes read
+    fd_set set;
+    struct timeval timeout{};
+    int len_to_read = _length;
+    char* read_ptr = static_cast<char*>( _buffer );
     _bytes_read = 0;
 
     while ( len_to_read > 0 ) {
         if ( nullptr != _time_value ) {
-            const int status = select( _socket + 1, &set, NULL, NULL, &timeout );
+            // Must always reset the fd_set and timeout before a call to select().
+            FD_ZERO(&set); // NOLINT
+            FD_SET(_socket, &set); // NOLINT
+            timeout = *_time_value;
+
+            log_net::debug("{}: Calling select() with timeout [{}.{}].", __func__, timeout.tv_sec, timeout.tv_usec);
+            const int status = select( _socket + 1, &set, nullptr, nullptr, &timeout );
+
             if ( status == 0 ) { // the select has timed out
-                return ERROR( SYS_SOCK_READ_TIMEDOUT, boost::format("socket timeout with [%d] bytes read") % _bytes_read);
-            } else if ( status < 0 ) {
+                log_net::debug("{}: select() timed out. timeout object now holds [{}.{}].", __func__, timeout.tv_sec, timeout.tv_usec);
+                return ERROR( SYS_SOCK_READ_TIMEDOUT, fmt::format("socket timeout with [{}] bytes read", _bytes_read));
+            }
+
+            if ( status < 0 ) {
+                log_net::debug("{}: select() encountered an error [{}], errno = [{}].", __func__, status, errno);
+
                 if ( errno == EINTR ) {
-                    continue;
-                } else {
-                    return ERROR( SYS_SOCK_READ_ERR - errno, boost::format("error on select after [%d] bytes read") % _bytes_read);
+                    // TODO Need a way to detect that SIGUSR1 signal was received.
+                    // Perhaps we read the value of g_terminate or simply return and handle
+                    // termination at an earlier place in the callstack.
+                    return ERROR(INTERRUPT_DETECTED, fmt::format("{} interrupted by signal", __func__));
```

Comment:
Handle TODO.

Need to think about whether this is still true.

Reply:
This is correct.

To maintain iRODS 4 behavior, I added a `continue` instruction in an earlier part of the stack. A comment is included in the code explaining why. See rsApiHandler.cpp for details.

Done.